devicetypes/shinasys/sihas-zigbee-metering-plug.src/sihas-zigbee-metering-plug.groovy

```diff
@@ -0,0 +1,175 @@
+	}
+}
+
+def getATTRIBUTE_READING_INFO_SET() { 0x0000 }
+def getATTRIBUTE_HISTORICAL_CONSUMPTION() { 0x0400 }
+def getATTRIBUTE_ACTIVE_POWER() { 0x050B }
+def getATTRIBUTE_FREQUENCY() { 0x0300 }
+def getATTRIBUTE_VOLTAGE() { 0x0505 }
+def getATTRIBUTE_CURRENT() { 0x0508 }
+def getATTRIBUTE_POWERFACTOR() { 0x0510 }
+def getTEMPERATURE_MEASUREMENT_MEASURED_VALUE_ATTRIBUTE() { 0x0000 }
+
+def convertHexToInt24Bit(value) {
+	int result = zigbee.convertHexToInt(value)
+	if (result & 0x800000) {
+		result |= 0xFF000000
+	}
+	return result
+}
+
+def parse(String description) {
+	log.debug "description is $description"
+	def event = zigbee.getEvent(description)
+	def descMap = zigbee.parseDescriptionAsMap(description)
+
+	if (event) {
+		log.info "event enter:$event"
+		if (event.name == "switch") {
+			return sendEvent(event)
+		} else if (event.name == "temperature") {
+			return sendEvent(event)
+		}
+	}
+
+	if (descMap) {
+		List result = []
+		log.debug "Desc Map: $descMap"
+
+		List attrData = [[clusterInt: descMap.clusterInt ,attrInt: descMap.attrInt, value: descMap.value, isValidForDataType: descMap.isValidForDataType]]
+		descMap.additionalAttrs.each {
+			attrData << [clusterInt: descMap.clusterInt, attrInt: it.attrInt, value: it.value, isValidForDataType: it.isValidForDataType]
+		}
+		attrData.each {
+			def map = [:]
+			if (it.isValidForDataType && (it.value != null)) {
+				if (it.clusterInt == zigbee.SIMPLE_METERING_CLUSTER && it.attrInt == ATTRIBUTE_HISTORICAL_CONSUMPTION) {
+					log.debug "meter"
+					map.name = "power"
+					map.value = convertHexToInt24Bit(it.value)/powerDivisor
+					map.unit = "W"
+				} else if (it.clusterInt == zigbee.SIMPLE_METERING_CLUSTER && it.attrInt == ATTRIBUTE_READING_INFO_SET) {
+					log.debug "energy"
+					map.name = "energy"
+					map.value = zigbee.convertHexToInt(it.value)/energyDivisor
+					map.unit = "kWh"						
+				} else if (it.clusterInt == zigbee.ELECTRICAL_MEASUREMENT_CLUSTER && it.attrInt == ATTRIBUTE_FREQUENCY) {
+					log.debug "frequency"
+					map.name = "frequency"
+					map.value = zigbee.convertHexToInt(it.value)/frequencyDivisor
+					map.unit = "Hz"
+				} else if (it.clusterInt == zigbee.ELECTRICAL_MEASUREMENT_CLUSTER && it.attrInt == ATTRIBUTE_VOLTAGE) {
+					log.debug "voltage"
+					map.name = "voltage"
+					map.value = zigbee.convertHexToInt(it.value)/voltageDivisor
+					map.unit = "V"
+				} else if (it.clusterInt == zigbee.ELECTRICAL_MEASUREMENT_CLUSTER && it.attrInt == ATTRIBUTE_CURRENT) {
+					log.debug "current"
+					map.name = "current"
+					map.value = zigbee.convertHexToInt(it.value)/currentDivisor
+					map.unit = "A"
+				} else if (it.clusterInt == zigbee.ELECTRICAL_MEASUREMENT_CLUSTER && it.attrInt == ATTRIBUTE_POWERFACTOR) {
+					log.debug "power factor"
+					map.name = "powerFactor"
+					map.value = (byte) zigbee.convertHexToInt(it.value)/powerFactorDivisor
```

Comment:
Why this value is casted to byte?

Reply:
![powerfactor_zcl](https://user-images.githubusercontent.com/71238736/152893593-6afdd7df-42bd-4f57-a781-8e116c8c65c6.jpg)
When the current transformer (CT) is input in the reverse direction, it comes up as a negative value.